File: embed.m
% Watermark Embedding

clear all; 

image_name='images/lena128x128.png';
watermark_name='images/baboon128x128.png';
output_name='outputs/watermarked.png';

% read in the images
image_object=imread(image_name);
watermark_object=imread(watermark_name);

% determine size of watermark image
Mw=size(watermark_object,1);    %Height
Nw=size(watermark_object,2);    %Width

% use msb of watermark image
for ii = 1:Mw
    for jj = 1:Nw
        watermark(ii,jj)=bitget(watermark_object(ii,jj),8);
    end
end

% set the lsb of image_object(ii,jj) to the value of the msb of watermark(ii,jj)
watermarked_object=image_object;
for ii = 1:Mw
    for jj = 1:Nw
        watermarked_object(ii,jj)=bitset(watermarked_object(ii,jj),1,watermark(ii,jj));
    end
end

% write to file the watermarked image
imwrite(watermarked_object,output_name);

File: embed.r
require("MMSec")

# Read original image
#data(lena)
lena <- load.png("images/lena-small.png")
original <- as.matrix(lena) * 255


# Size of original image
dim <- dim(original)
heigth <- dim[1]
width <- dim[2]

# Generate watermark
watermark <- matrix(0, nrow=heigth, ncol=width)

for (i in 1:heigth) {
	watermark[i,] <- sample(0:255, width, replace=T)
}

# Store the watermark
write.table(watermark, file="outputs/watermark", row.names=F, col.names=F)

save.png(greymap(watermark), "outputs/watermark.png", overwrite=TRUE)

# Apply λ to the watermark
λ <- 1/51
watermark <- λ * watermark
watermark <- round(watermark)

# Adding watermark to the original image
watermarked <- original

for (i in 1:heigth) {
	for (j in 1:width) {
	
		watermarked[i,j] = original[i,j] + watermark[i,j]

		if (watermarked[i,j] > 255) {
			watermarked[i,j] = 255
		}

	}
}

# Save images
save.png(greymap(original), "outputs/original.png", overwrite=TRUE)
save.png(greymap(watermarked), "outputs/watermarked.png", overwrite=TRUE)

